### PR TITLE
Full recurse and convert all dicts beneath a BTree to a BTree

### LINKED ISSUE #41: BTree conversion: Wide but shallow _adapters can lose data in ZODB
If the top-level becomes a BTree, but the inner level (`{name: utility}`) is still small enough to be a `dict`, data can be lost. This is because the inner level is mutated in place, and because it's a value of the BTree, the BTree (or its bucket) doesn't get `_p_changed` set so it doesn't know to register with the connection to be written out.

The simple solution to this is to set the conversion threshold to 0, but perhaps we can do better. Especially because while that should work for utility registrations, it won't work for complicated adapter registrations that go many layers deep.

### PR DESCRIPTION
**Pull request by jamadden**
This solves the `_p_changed` issue and fixes #41 

In practice, this will be slightly slower for mutating large utility registrations, but probably not by much (I haven't tried to measure).

### CODE REVIEW

**Comment by jamadden** [comment]
I've also added a method to synchronize the known-good data in `_utility_registrations` with the data in `.utilities`.

File: src/nti/site/site.py [hunk 1]
```diff
@@ -314,51 +315,108 @@ def _btlar_after_setstate(self):
...
+    def __mapping_needs_converted(self, path, mapping, parent_is_btree, btree_type):
+        # We need to convert if the parent is a btree (and we're not a btree already)
+        # OR we're not a btree and we're too long
+        regs_is_btree = isinstance(mapping, btree_type)
+
+        needs_converted = (
+            parent_is_btree and not regs_is_btree
+        ) or (
+            not regs_is_btree and len(mapping) > self.btree_map_threshold
+        )
```
Comment (jzuech3) [inline src/nti/site/site.py:335]: Worth simplifying this for readability?

```suggestion
        needs_converted=False
        if not regs_is_btree:
            needs_converted = parent_is_btree or len(mapping) > self.btree_map_threshold
```

File: src/nti/site/site.py [hunk 2]
```diff
@@ -314,51 +315,108 @@ def _btlar_after_setstate(self):
...
+            for k, v in mapping.items():
+                new_v = self.__btree_map(
+                    path + (k,),
+                    v,
+                    depth + 1,
+                    max_depth,
+                    parent_is_btree,
+                    btree_type
+                )
+                if new_v is not v:
```
Comment (jzuech3) [inline src/nti/site/site.py:370]: Is `is not` the correct check here? Is that checking if the instance is the same?

File: src/nti/site/site.py (hunk 1 above)
Comment (jamadden) [reply src/nti/site/site.py:335]: I find that less readable 😢

Comment (cutz) [reply src/nti/site/site.py:335]: I agree, I like the current form of the conditional.

File: src/nti/site/site.py (hunk 2 above)
Comment (jamadden) [reply src/nti/site/site.py:370]: Correct. If we mutated something in place, there's no need to keep holding on to it here.

I had a more complicated version of this that tracked if anything anywhere beneath something changed and rewrote the tree all the way up, but that was complex and turned out to be unnecessary if I started conversion from the bottom but tracked if a parent was (or became) a btree. The `parent_is_btree` argument simplified everything a bunch.

File: src/nti/site/site.py [hunk 3]
```diff
@@ -314,51 +315,108 @@ def _btlar_after_setstate(self):
...
+        # type: (tuple, dict, int, int, bool, type) -> dict
+        #
+        # path is just for information
+        if self.__mapping_needs_converted(
+                path, mapping, parent_is_btree, btree_type
+        ):
+            mapping = btree_type(mapping)
+            # Force conversion of child leaves
+            parent_is_btree = True
+            #changed_below = True
```
Comment (jzuech3) [inline src/nti/site/site.py:354]: Leftover logic

**Review by jzuech3 — APPROVED** [review]
LGTM

**Pull request closed without merging**